[bug] CMakeDeps errors out when a package is both "tool_requires" & "requires"

### Environment details

* Operating System+version: Rocky Linux 8.8
* Compiler+version: GCC 8.5.0
* Conan version: 2.0.11
* Python version: 3.9.16


### Steps to reproduce

1. Please, clone this test repo : https://github.com/Todiq/test_conan.git
2. Run `python3 build.py`

### Logs

```bash
Reference 'alpha/1.0' in editable mode

======== Input profiles ========
Profile host:
[settings]
arch=x86_64
build_type=Release
compiler=gcc
compiler.cppstd=gnu17
compiler.libcxx=libstdc++11
compiler.version=8
os=Linux

Profile build:
[settings]
arch=x86_64
build_type=Release
compiler=gcc
compiler.cppstd=gnu17
compiler.libcxx=libstdc++11
compiler.version=8
os=Linux


======== Computing dependency graph ========
Graph root
    conanfile.py (alpha/1.0): /home/carl/test_conan/Test/alpha/conanfile.py
Requirements
    zlib/1.2.13#97d5730b529b4224045fe7090592d4c1 - Cache

======== Computing necessary packages ========
Requirements
    zlib/1.2.13#97d5730b529b4224045fe7090592d4c1:f940c1f7889f9018080dbeebdede5a67d6c3b30e#f6c432466a22a8650ef0e6b459b4722e - Cache

======== Installing packages ========

======== Installing packages ========
zlib/1.2.13: Already installed! (1 of 1)
WARN: deprecated: Usage of deprecated Conan 1.X features that will be removed in Conan 2.X:
WARN: deprecated:     'cpp_info.names' used in: zlib/1.2.13

======== Finalizing install (deploy, generators) ========
conanfile.py (alpha/1.0): Calling generate()
conanfile.py (alpha/1.0): Generators folder: /home/carl/test_conan/Test/alpha/build/Release/generators
conanfile.py (alpha/1.0): CMakeToolchain generated: conan_toolchain.cmake
conanfile.py (alpha/1.0): Preset 'conan-release' added to CMakePresets.json. Invoke it manually using 'cmake --preset conan-release' if using CMake>=3.23
conanfile.py (alpha/1.0): If your CMake version is not compatible with CMakePresets (<3.23) call cmake like: 'cmake <path> -G "Unix Makefiles" -DCMAKE_TOOLCHAIN_FILE=/home/carl/test_conan/Test/alpha/build/Release/generators/conan_toolchain.cmake -DCMAKE_POLICY_DEFAULT_CMP0091=NEW -DCMAKE_BUILD_TYPE=Release'
conanfile.py (alpha/1.0): CMakeToolchain generated: CMakePresets.json
conanfile.py (alpha/1.0): CMakeToolchain generated: ../../../CMakeUserPresets.json
conanfile.py (alpha/1.0): Generating aggregated env files
conanfile.py (alpha/1.0): Generated aggregated env files: ['conanbuild.sh', 'conanrun.sh']

======== Calling build() ========
conanfile.py (alpha/1.0): Calling build()
conanfile.py (alpha/1.0): Running CMake.configure()
conanfile.py (alpha/1.0): RUN: cmake -G "Unix Makefiles" -DCMAKE_TOOLCHAIN_FILE="/home/carl/test_conan/Test/alpha/build/Release/generators/conan_toolchain.cmake" -DCMAKE_INSTALL_PREFIX="/home/carl/test_conan/Test/alpha" -DCMAKE_POLICY_DEFAULT_CMP0091="NEW" -DCMAKE_BUILD_TYPE="Release" "/home/carl/test_conan/Test/alpha"
-- Using Conan toolchain: /home/carl/test_conan/Test/alpha/build/Release/generators/conan_toolchain.cmake
-- Conan toolchain: C++ Standard 17 with extensions ON
-- Conan toolchain: Setting BUILD_SHARED_LIBS = ON
-- The CXX compiler identification is GNU 8.5.0
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /usr/bin/c++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Conan: Target declared 'ZLIB::ZLIB'
-- Configuring done (0.2s)
-- Generating done (0.0s)
CMake Warning:
  Manually-specified variables were not used by the project:

    CMAKE_POLICY_DEFAULT_CMP0091


-- Build files have been written to: /home/carl/test_conan/Test/alpha/build/Release

conanfile.py (alpha/1.0): Running CMake.build()
conanfile.py (alpha/1.0): RUN: cmake --build "/home/carl/test_conan/Test/alpha/build/Release" -- -j8
[ 16%] Building CXX object alpha1/CMakeFiles/alpha1.dir/src/alpha1.cpp.o
[ 33%] Linking CXX shared library libalpha1.so
[ 33%] Built target alpha1
[ 50%] Building CXX object alpha2/CMakeFiles/alpha2.dir/src/alpha2.cpp.o
[ 66%] Linking CXX shared library libalpha2.so
[ 66%] Built target alpha2
[ 83%] Building CXX object alpha_exe/CMakeFiles/alpha_exe.dir/src/alpha_exe.cpp.o
[100%] Linking CXX executable alpha_exe
[100%] Built target alpha_exe


======== Input profiles ========
Profile host:
[settings]
arch=x86_64
build_type=Release
compiler=gcc
compiler.cppstd=gnu17
compiler.libcxx=libstdc++11
compiler.version=8
os=Linux

Profile build:
[settings]
arch=x86_64
build_type=Release
compiler=gcc
compiler.cppstd=gnu17
compiler.libcxx=libstdc++11
compiler.version=8
os=Linux


======== Launching test_package ========

======== Computing dependency graph ========
Graph root
    alpha/1.0 (test package): /home/carl/test_conan/Test/alpha/test_package/conanfile.py
Requirements
    alpha/1.0 - Editable
    zlib/1.2.13#97d5730b529b4224045fe7090592d4c1 - Cache
Build requirements
    alpha/1.0 - Editable
    zlib/1.2.13#97d5730b529b4224045fe7090592d4c1 - Cache

======== Computing necessary packages ========
Requirements
    alpha/1.0:04902910cc170dfbfd8536063e51c0ee2e58fb32 - Editable
    zlib/1.2.13#97d5730b529b4224045fe7090592d4c1:f940c1f7889f9018080dbeebdede5a67d6c3b30e#f6c432466a22a8650ef0e6b459b4722e - Cache
Build requirements
    alpha/1.0:04902910cc170dfbfd8536063e51c0ee2e58fb32 - Editable
    zlib/1.2.13#97d5730b529b4224045fe7090592d4c1:f940c1f7889f9018080dbeebdede5a67d6c3b30e#f6c432466a22a8650ef0e6b459b4722e - Cache

======== Installing packages ========
zlib/1.2.13: Already installed! (1 of 2)
alpha/1.0: Rewriting files of editable package 'alpha' at '/home/carl/test_conan/Test/alpha/build/Release/generators'
alpha/1.0: Calling generate()
alpha/1.0: Generators folder: /home/carl/test_conan/Test/alpha/build/Release/generators
alpha/1.0: CMakeToolchain generated: conan_toolchain.cmake
alpha/1.0: CMakeToolchain generated: CMakePresets.json
alpha/1.0: CMakeToolchain generated: ../../../CMakeUserPresets.json
alpha/1.0: Generating aggregated env files
alpha/1.0: Generated aggregated env files: ['conanbuild.sh', 'conanrun.sh']
WARN: deprecated: Usage of deprecated Conan 1.X features that will be removed in Conan 2.X:
WARN: deprecated:     'cpp_info.names' used in: zlib/1.2.13

======== Testing the package ========
alpha/1.0 (test package): Test package build:
alpha/1.0 (test package): Test package build folder: /home/carl/test_conan/Test/alpha/test_package
alpha/1.0 (test package): Writing generators to /home/carl/test_conan/Test/alpha/test_package
alpha/1.0 (test package): Generator 'CMakeDeps' calling 'generate()'
alpha/1.0 (test package): ERROR: Traceback (most recent call last):
  File "/usr/local/lib/python3.9/site-packages/conan/tools/cmake/cmakedeps/templates/__init__.py", line 35, in render
    context = self.context
  File "/usr/local/lib/python3.9/site-packages/conan/tools/cmake/cmakedeps/templates/target_data.py", line 38, in context
    global_cpp = self._get_global_cpp_cmake()
  File "/usr/local/lib/python3.9/site-packages/conan/tools/cmake/cmakedeps/templates/target_data.py", line 185, in _get_global_cpp_cmake
    return _TargetDataContext(global_cppinfo, pfolder_var_name, self._root_folder,
  File "/usr/local/lib/python3.9/site-packages/conan/tools/cmake/cmakedeps/templates/target_data.py", line 298, in __init__
    self.include_paths = join_paths(cpp_info.includedirs)
  File "/usr/local/lib/python3.9/site-packages/conan/tools/cmake/cmakedeps/templates/target_data.py", line 272, in join_paths
    assert os.path.isabs(p), "{} is not absolute".format(p)
AssertionError: include is not absolute

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/lib/python3.9/site-packages/conans/client/generators/__init__.py", line 94, in write_generators
    generator.generate()
  File "/usr/local/lib/python3.9/site-packages/conan/tools/cmake/cmakedeps/cmakedeps.py", line 43, in generate
    generator_files = self.content
  File "/usr/local/lib/python3.9/site-packages/conan/tools/cmake/cmakedeps/cmakedeps.py", line 85, in content
    self._generate_files(require, dep, ret, find_module_mode=False)
  File "/usr/local/lib/python3.9/site-packages/conan/tools/cmake/cmakedeps/cmakedeps.py", line 98, in _generate_files
    data_content = relativize_generated_file(data_target.render(), self._conanfile,
  File "/usr/local/lib/python3.9/site-packages/conan/tools/cmake/cmakedeps/templates/__init__.py", line 37, in render
    raise ConanException("error generating context for '{}': {}".format(self.conanfile, e))
conans.errors.ConanException: error generating context for 'alpha/1.0': include is not absolute

ERROR: Error in generator 'CMakeDeps': error generating context for 'alpha/1.0': include is not absolute
Traceback (most recent call last):
  File "/home/carl/test_conan/build.py", line 17, in <module>
    run('conan test Test/alpha/test_package alpha/1.0')
  File "/home/carl/test_conan/build.py", line 6, in run
    raise Exception(f"Failed CMD: {cmd}")
Exception: Failed CMD: conan test Test/alpha/test_package alpha/1.0
```

--- a/conans/model/build_info.py
+++ b/conans/model/build_info.py
@@ -459,7 +459,8 @@ class CppInfo:
         self._package.sysroot = self._package.sysroot or other._package.sysroot
         # COMPONENTS
         for cname, c in other.components.items():
-            self.components[cname].merge(c, overwrite)
+            # Make sure each component created on the fly does not bring new defaults
+            self.components.setdefault(cname, _Component(set_defaults=False)).merge(c, overwrite)
 
     def set_relative_base_folder(self, folder):
         """Prepend the folder to all the directories definitions, that are relative"""

--- a/conans/test/integration/editable/test_editables_layout.py
+++ b/conans/test/integration/editable/test_editables_layout.py
@@ -277,3 +277,43 @@ def test_editable_package_folder():
     c.run("editable add .")
     c.run("install --requires=pkg/0.1@")
     assert "pkg/0.1: PKG FOLDER=None!!!"
+
+
+def test_editable_components_absolute_paths():
+    """
+    this was failing in https://github.com/conan-io/conan/issues/14777
+    because components aggregation had a bug
+    """
+    c = TestClient()
+    conanfile = textwrap.dedent("""
+        from conan import ConanFile
+
+        class Pkg(ConanFile):
+            name = "alpha"
+            version = "1.0"
+
+            def layout(self):
+                self.cpp.source.components["headers"].includedirs = ["include222"]
+                self.cpp.build.components["alpha1"].libdirs = ["alpha1",]
+                self.cpp.build.components["alpha2"].libdirs = ["alpha2"]
+                self.cpp.build.components["alpha_exe"].bindirs = ["alpha_exe"]
+        """)
+    test = textwrap.dedent("""
+        from conan import ConanFile
+
+        class Pkg(ConanFile):
+            settings = "os", "compiler", "build_type", "arch"
+            generators = "CMakeDeps"
+
+            def requirements(self):
+                self.requires(self.tested_reference_str)
+
+            def test(self):
+                pass
+            """)
+    c.save({"conanfile.py": conanfile,
+            "test_package/conanfile.py": test})
+    c.run("editable add .")
+    c.run("create .")
+    # This used to crash due to "include" is not absolute path, now it works
+    assert "Testing the package" in c.out

